### PR TITLE
wg_engine: composition and blend optimization - PART 1

### PR DESCRIPTION
**Pull request by SergeyLebedkin**
Whats new
* bind groups creation in real time removed - performance boost
* blend and composition shaders decomposed - performance boost
* shader modules and pipeline layouts generalized - less memory usage
* shared single stencil buffer used - less memory usage
* bind groups usage simplified
* general context API simplified and generalized
* all rendering logic moved into new composition class
* ready for hardware MSAA (in next steps)
* ready for direct mask applience (in next steps)

web/emsdk verified

Examples to verify:
* MaskingMethods
* SceneBlending
* Blending
* ClipPath

Old (76 fps)/New (160 fps)
image

Shader performance (old/new) (more than 2 times):
image

### CODE REVIEW

**Comment by tinyjin** [comment]
@SergeyLebedkin 
I've tested the animation in the browser. For the masking sample, I see similar FPS to previous build. Test result still indicates about 41FPS.

image

**Comment by SergeyLebedkin** [comment]
> @SergeyLebedkin I've tested the animation in the browser. For the masking sample, I see similar FPS to previous build. Test result still indicates about 41 FPS in my test environment.
> 
> * Full Size animation in 14-inch(3024 × 1964)

FPS should we much better, but in current state thats OK. Main reason for testing this PR is to check that every thing workins correctly and stable in web environment

File: src/renderer/wg_engine/tvgWgRenderData.cpp [hunk 1]
```diff
@@ -261,11 +259,25 @@ void WgRenderSettings::release(WgContext& context)
...
+    context.releaseBuffer(bufferBlendSettings);
     clips.clear();
 };
 
 
+void WgRenderDataPaint::update(WgContext& context, const tvg::Matrix& transform, tvg::ColorSpace cs, uint8_t opacity)
+{
+    WgShaderTypeMat4x4f modelMat(transform);
+    WgShaderTypeBlendSettings blendSettings(cs, opacity);
+    if ((context.allocateBufferUniform(bufferModelMat, &modelMat, sizeof(modelMat))) |
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgRenderData.cpp:273]: `|` -> `||` ?

**Review by mgrudzinska — CHANGES_REQUESTED** [review]
Lottie example & thread sanitizer - crash
Lottie example & asan - errors:

1. Validation Error

Caused by:
    In wgpuDeviceCreateBindGroup
    The adapter does not support read access for storages texture of format Bgra8Unorm

2. Validation Error

Caused by:
    In wgpuRenderPassEncoderEnd
      note: encoder = `<CommandBuffer-(0, 1, Metal)>`
    In a pass parameter
      note: command buffer = `<CommandBuffer-(0, 1, Metal)>`
    Command encoder is invalid

3. Validation Error

Caused by:
    In wgpuCommandEncoderCopyTextureToTexture
    Command encoder is invalid

File: src/renderer/wg_engine/tvgWgRenderData.cpp (hunk 1 above)
Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgRenderData.cpp:273]: No, MUST be | instead of ||

Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgRenderData.cpp:273]: because you want to ensure both are called? can't you use && ?

Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgRenderData.cpp:273]: now we get warning, if it has to be `|` cast into int

Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgRenderData.cpp:273]: We need to chick if any of _context.allocateBufferUniform_ return false

Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgRenderData.cpp:273]: OK... I think I know how to fix warning properly. And this worning appears only on clang compiler. I will rewrite this logic properly

File: src/renderer/wg_engine/tvgWgCompositor.cpp [hunk 2]
```diff
@@ -0,0 +1,509 @@
+/*
+ * Copyright (c) 2023 - 2024 the ThorVG project. All rights reserved.
```
[the remark this answers is not in the document]
Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgCompositor.cpp:2]: @hermet Please, can you resolve this question?

File: src/renderer/wg_engine/tvgWgCompositor.cpp [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+        drawImage(context, renderData, blentType);
+};
+
+
+void WgCompositor::blendShape(WgContext& context, WgRenderDataShape* renderData, BlendMethod blendMethod)
+{
+    assert(renderData);
+    assert(commandEncoder);
+    assert(currentTarget);
+    // skip shape blending if they are not exists
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgCompositor.cpp:198]: skip shape blending if they do not exist

File: src/renderer/wg_engine/tvgWgCompositor.cpp [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+};
+
+
+void WgCompositor::composeShape(WgContext& context, WgRenderDataShape* renderData, WgRenderStorage* msk, CompositeMethod composeMethod)
+{
+    assert(msk);
+    assert(renderData);
+    assert(commandEncoder);
+    assert(currentTarget);
+    // skip shapes composing if shapes are not exists
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgCompositor.cpp:261]: skip shapes blending if they do not exist

File: src/renderer/wg_engine/tvgWgCompositor.cpp [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+    // restore current render pass
+    beginRenderPass(commandEncoder, target, false);
+}
+
+void WgCompositor::blendStrokes(WgContext& context, WgRenderDataShape* renderData, BlendMethod blendMethod)
+{
+    assert(renderData);
+    assert(commandEncoder);
+    assert(currentTarget);
+    // skip strokes blending if they are not exists
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgCompositor.cpp:219]: 'do not exist' and in other places as well

File: src/renderer/wg_engine/tvgWgCompositor.cpp [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+    // run blend
+    blend(commandEncoder, &storageInterm, target, 255, blendMethod, WgRenderRasterType::Image);
+    // restore current render pass
+    beginRenderPass(commandEncoder, target, false);
+};
+
+
+void WgCompositor::composeShape(WgContext& context, WgRenderDataShape* renderData, WgRenderStorage* msk, CompositeMethod composeMethod)
+{
+    assert(msk);
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgCompositor.cpp:257]: can we use a name 'mask' instead 'msk' ?

File: src/renderer/wg_engine/tvgWgCompositor.cpp (hunk 4 above)
Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgCompositor.cpp:261]: in general, not only in this place

File: src/renderer/wg_engine/tvgWgCompositor.cpp (hunk 6 above)
Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgCompositor.cpp:257]: its the same name as in the shaders, but yes, we can

File: src/renderer/wg_engine/tvgWgCompositor.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2023 - 2024 the ThorVG project. All rights reserved.
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgCompositor.h:2]: 2024

File: src/renderer/wg_engine/tvgWgCompositor.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    // global stencil buffer handles
+    WGPUTexture texStencil{};
+    WGPUTextureView texViewStencil{};
+    // global view matrix handles
+    WGPUBuffer bufferViewMat{};
+    WGPUBindGroup bindGroupViewMat{};
+    // opacity value pool
+    WGPUBuffer bufferOpacities[256]{};
+    WGPUBindGroup bindGroupOpacities[256]{};
+    // current render pass handes
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgCompositor.h:46]: handles

File: src/renderer/wg_engine/tvgWgPipelines.cpp [hunk 9]
```diff
@@ -23,492 +23,359 @@
...
+            primitiveState, multisampleState, blendStates[i]);
+    }
 
-//************************************************************************
-// pipelines
-//************************************************************************
+    // render pipeline linear
+    for (uint32_t i = 0; i < 3; i++) {
+        linear[i] = createRenderPipeline(
+            context.device, "The render pipeline linear",
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgPipelines.cpp:283]: the render?

Comment (mgrudzinska) [reply src/renderer/wg_engine/tvgWgPipelines.cpp:283]: same question to all other pipelines

File: src/renderer/wg_engine/tvgWgPipelines.cpp [hunk 10]
```diff
@@ -23,492 +23,359 @@
...
+    releaseShaderModule(shaderImage);
+    releaseShaderModule(shaderLinear);
+    releaseShaderModule(shaderRadial);
+    releaseShaderModule(shaderSolid);
+    releaseShaderModule(shaderStencil);
 }
 
 
-bool WgPipelines::isBlendMethodSupportsHW(BlendMethod blendMethod)
+void WgPipelines::releaseComputes(WgContext& context)
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgPipelines.cpp:350]: Compute resources? computations?

**Review by mgrudzinska — APPROVED** [review]

**Pull request merged**